src/java.net.http/share/classes/java/net/http/HttpResponse.java

```diff
@@ -143,6 +144,14 @@ public interface HttpResponse<T> {
      */
     public T body();
 
+    /**
+     * Returns the body if the given predicate is satisfied.
+     *
+     * @param predicate the Predicate to test
+     * @return an Optional containing the response body if the predicate returns true
+     */
+    public Optional<T> bodyWhen(Predicate<ResponseInfo> predicate) throws IOException;
+
```

Comment:
I believe the method should be specified to close the body if it is closeable and if it is not returned.

Reply:
I wasn't sure about that, because that effectively means this method can be called once only (in the case closeable response bodies). But, maybe that is not unreasonable in those cases. I'll change as suggested.